### ISSUE TITLE
 almost weekly useful materials - 01/31 -

### ISSUE THREAD

**Issue by GENZITSU** [GENZITSU/UsefulMaterials#144]

**Comment by GENZITSU** [comment]
## Fine-Tuning LLMs: LoRA or Full-Parameter? An in-depth Analysis with Llama 2

Full Parameter学習とLoRA学習の性能差や学習効率などについて述べられているブログ。

LoRAの基礎的な仕組みから説明がなされているため勉強になる。

以下勉強になった点を抜粋

### パフォーマンス比較

タスクにも寄るが、7B, 13B, 70BでLoRAとfull parametersはほとんど性能差が無い

![anyscale-all](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/7357c508-065c-4b64-853b-cd8fd0b68874)


### LoRAのハイパラ

rank(r): 8~16くらいで十分だったとのこと、当然8の方がcheckpointのサイズは小さくなる

lora_alpha: 16で固定するのが良いとされている。 ([元論文](https://arxiv.org/abs/2106.09685), [Platypusの論文](https://arxiv.org/pdf/2308.07317v1.pdf))

Target modules: 全パラメータを対象にした方が良い

base learning rate: 1e-4がよく使われているが、もっと低い方が安定する。1e-5や3e-5など

![スクリーンショット 2024-01-24 13 00 01](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/07169251-8c95-4c6e-be24-8f53e5a6d9b0)

![スクリーンショット 2024-01-24 13 00 11](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/399c4a71-01e3-41f1-89d5-fdc1642d6ad8)

### 学習時に用いるプロンプト

full finetuningする際は特殊トークンを用いたプロンプトテンプレートの利用が効果的だったが、LoRAだとあまり効果が出ないとのこと(out-of-distributionすぎるから...?)

![lora_blog_code_1](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/ffeb1c8f-0faa-47a4-a50e-c501238db3b9)

以下のようにタスク内容を詳細に説明してあげるようなプロンプトにすると学習が効率的に進んだとのこと

![lora_blog_code_2](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/7b0585e3-4dce-49eb-9ac5-9e3635d3a86f)

![スクリーンショット 2024-01-24 13 04 50](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/f3c1d6a4-07db-416f-8a4a-1057ac670d03)

### LoRA時の特殊トークンの扱い方

追加したトークンの分だけ追加のembedding matrixを作成し、Loraをかける。
推論時に追加したembedding matrixもcheckpointとして保存しておくことを忘れずに

![スクリーンショット 2024-01-24 13 15 24](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/a7abd1ad-9027-461a-ae35-7b0629b40470)

### 学習速度について

1系列あたりのメモリ消費が効率的になった分だけバッチサイズを上げないと学習のスループット速度は上がらないことに注意。

![スクリーンショット 2024-01-24 13 17 55](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/81f548e2-9968-4b1b-ad43-b21efdc032e1)

一方で特定性能に達するまでの時間という意味だとfull parameterとLoRAであまり差はない

![スクリーンショット 2024-01-24 13 18 56](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/5b837244-e699-4c3e-82ea-bb951fec0952)

### メモリ消費量

概ねLoRA(r=8)の場合半分程度の消費になる。

![スクリーンショット 2024-01-24 13 22 20](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/7f721e7b-d3a0-4bef-af90-52ff9c1c34a4)

### モデルの保存容量

![スクリーンショット 2024-01-24 13 22 59](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/24e55b4a-b37a-45c0-bbc6-e9236e0f9d5f)

Lora8の全レイヤー保存の場合でも1/350程度になる。モデルを複数サービングするときに効果が大きく出る

### コメント

学習スループットは上がるが、学習完了までの時間は結局変わらないというのが意外だった。
パラメータ設定部分も参考になる。

### 出典
- [Fine-Tuning LLMs: LoRA or Full-Parameter? An in-depth Analysis with Llama 2](https://www.anyscale.com/blog/fine-tuning-llms-lora-or-full-parameter-an-in-depth-analysis-with-llama-2#lora-vs-full-parameter-fine-tuning:-things-to-consider)

**Comment by GENZITSU** [comment]
## [brevdev](https://github.com/brevdev)/ [notebooks](https://github.com/brevdev/notebooks)

LLMの推論や学習方法を記載したnotebookをまとめているレポジトリ

![スクリーンショット 2024-01-24 14 37 04](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/4019b865-2c50-44a8-9fc3-8ea72d762e45)

### コメント

いつか参考にできそう

### 出典
-  [brevdev](https://github.com/brevdev)/ [notebooks](https://github.com/brevdev/notebooks)

**Comment by GENZITSU** [comment]
## airllmを使ってT4で70B LLMの推論を実行してみる

70Bなどのモデルサイズがでかいモデルを小メモリPCでも動かせるように層ごとに推論させるライブラリairllmの紹介

t4インスタンスで`meta-llama/Llama-2-70b-chat-hf1`を動かしたときに1 tokenあたり 7~8分かかったようで実用はなかなか難しそう。

ただし性能は結構良いようで、AWPなどを駆使したMLMモデルのaccが0.814だったものが0.84程度になったとのこと
(厳密には評価データが異なる)

### コメント
もう少し最適化されて1分くらいになれば使い道ありそう

### 出典
- [airllmを使ってT4で70B LLMの推論を実行してみる](https://www.ai-shift.co.jp/techblog/4162)

**Comment by GENZITSU** [comment]
## Yes, Transformers are Effective for Time Series Forecasting (+ Autoformer)

AAAI2023にて投げかけられた[Are Transformers Effective for Time Series Forecasting?](https://arxiv.org/abs/2205.13504) という問いに対して、時系列データに対してもTransformer系のモデルは効果的であるという結果を実験的に示している記事

以下が結果

1次元の時系列データに対しての方が多次元のものよりも常によく、Transformerでも全然精度が出ることを示している。

![スクリーンショット 2024-01-26 18 33 01](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/4e2bc2a8-b5c6-4012-9265-b362b8232d52)

ただし、学習データのサイズが十分にないと効力が発揮しづらいとも述べられているため、実際のデータに適用する時はよく考えないとダメそう。

### コメント

評価指標的にはTransofmerはいいぞって感じになっているが、予測結果のplotを見るとほんまか...?ってなる

まぁ時系列予測自体が難しいというのが大きそう

![スクリーンショット 2024-01-26 18 39 02](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/9c2b1444-73e3-492e-a4b7-04990fe4ec6a)
![スクリーンショット 2024-01-26 18 38 55](https://github.com/GENZITSU/UsefulMaterials/assets/19774756/8bd6e07d-d57d-4726-a9b9-eee91ba6502f)


### 出典
- [Yes, Transformers are Effective for Time Series Forecasting (+ Autoformer)](https://huggingface.co/blog/autoformer)
